hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/MoveManager.java

```diff
@@ -0,0 +1,490 @@
+    // the container
+    DELETE_FAIL_POLICY,
+    //  replicas + target - src does not satisfy placement policy
+    REPLICATION_NOT_HEALTHY,
+    //write DB error
+    FAIL_CAN_NOT_RECORD_TO_DB
+  }
+
+  public static final Logger LOG =
+      LoggerFactory.getLogger(MoveManager.class);
+
+  // TODO - Should pending ops notify under lock to allow MM to schedule a
+  //        delete after the move, but before anything else can, eg RM?
+
+  // TODO - these need to be config defined somewhere, probably in the balancer
+  private static final long MOVE_DEADLINE = 1000 * 60 * 60; // 1 hour
+  private static final double MOVE_DEADLINE_FACTOR = 0.95;
+
+  private final ReplicationManager replicationManager;
+  private final ContainerManager containerManager;
+  private final Clock clock;
+
+  private final Map<ContainerID,
+      Pair<CompletableFuture<MoveResult>, MoveDataNodePair>> pendingMoves =
+      new ConcurrentHashMap<>();
+
+  private volatile boolean running = false;
+
+  public MoveManager(final ReplicationManager replicationManager,
+      final Clock clock, final ContainerManager containerManager) {
+    this.replicationManager = replicationManager;
+    this.containerManager = containerManager;
+    this.clock = clock;
+  }
+
+  /**
+   * get all the pending move operations.
+   */
+  public Map<ContainerID,
+      Pair<CompletableFuture<MoveResult>, MoveDataNodePair>> getPendingMove() {
+    return pendingMoves;
+  }
+
+  /**
+   * completeMove a move action for a given container.
+   *
+   * @param cid Container id to which the move option is finished
+   */
+  private void completeMove(final ContainerID cid, final MoveResult mr) {
+    Pair<CompletableFuture<MoveResult>, MoveDataNodePair> move =
+        pendingMoves.remove(cid);
+    if (move != null) {
+      CompletableFuture<MoveResult> future = move.getLeft();
+      if (future != null && mr != null) {
+        // when we know the future is null, and we want to complete
+        // the move , then we set mr to null.
+        // for example , after becoming a new leader, we might need
+        // to complete some moves and we know these futures does not
+        // exist.
+        future.complete(mr);
+      }
+    }
+  }
+
+  /**
+   * start a move for a given container.
+   *
+   * @param containerInfo The container for which to move a replica
+   * @param src source datanode to copy the replica from
+   * @param tgt target datanode to copy the replica to
+   */
+  private void startMove(
+      final ContainerInfo containerInfo, final DatanodeDetails src,
+      final DatanodeDetails tgt, final CompletableFuture<MoveResult> ret)
+      throws ContainerReplicaNotFoundException,
+      ContainerNotFoundException {
+    if (!pendingMoves.containsKey(containerInfo.containerID())) {
+      MoveDataNodePair mp = new MoveDataNodePair(src, tgt);
+      sendReplicateCommand(containerInfo, tgt, src);
+      pendingMoves.putIfAbsent(containerInfo.containerID(), Pair.of(ret, mp));
```

Comment:
Sorry I missed this earlier.  We should use an atomic operation instead of containsKey+putIfAbsent:

```suggestion
    pendingMoves.computeIfAbsent(containerInfo.containerID(), key -> {
      MoveDataNodePair mp = new MoveDataNodePair(src, tgt);
      sendReplicateCommand(containerInfo, tgt, src);
      return Pair.of(ret, mp));
    });
```

(Hope I got the syntax right.)

Reply:
yea makes sense. I had to change it a little to handle exceptions in the sendReplicateCommand, but the general ideal is the same.